read-fonts/src/tables/gsub/closure.rs

```diff
@@ -0,0 +1,188 @@
+//! Computing the closure over a set of glyphs
+//!
+//! This means taking a set of glyphs and finding the set of glyphs that are
+//! reachable from those glyphs via substitution.
+
+use std::collections::HashSet;
+
+use font_types::GlyphId;
+
+use crate::ReadError;
+
+use super::{
+    AlternateSubstFormat1, Gsub, LigatureSubstFormat1, MultipleSubstFormat1, SingleSubst,
+    SingleSubstFormat1, SingleSubstFormat2, SubstitutionSubtables,
+};
+
+#[cfg(feature = "std")]
+impl<'a> Gsub<'a> {
+    /// Return the set of glyphs reachable from the input set via any substituion.
+    pub fn closure_glyphs(&self, glyphs: HashSet<GlyphId>) -> Result<HashSet<GlyphId>, ReadError> {
+        // we need to do this iteratively, since any glyph found in one pass
+        // over the lookups could also be the target of substitutions.
+
+        let mut all_glyphs = glyphs;
+        let mut new_glyphs = HashSet::new();
+
+        // we always call this once, and then keep calling if it produces
+        // additional glyphs
+        self.closure_glyphs_once(&all_glyphs, &mut new_glyphs)?;
+        while !new_glyphs.is_subset(&all_glyphs) {
```

Comment:
Could we use a single hash set here and just iterate as long as the size changes? That seems equivalent to me and might require fewer passes because outputs immediately become inputs that later subtables can process.

Reply:
yes, fonttools also uses only one set and breaks the loop when it stops changing

Reply:
actually sorry no, fonttools still uses two sets, a mutable one that accummualtes the union of all the glyphs reachable so far (the output) and a frozenset that is an immutable copy of the former (acts as the input for the substitutions at each pass); the while True loop closing over all the lookups breaks when the ouput set does not change.
Distinguishing what goes in and out feels a bit clearer to me. But maybe Chad is right, taking all from the same cauldron and putting them back there as you co along could shorten the process and produce the same effect.
Though I'm not 100% sure about this.

Reply:
Update: yes it looks like this is no problem.